cucumber-core/src/main/java/io/cucumber/core/runner/CachingGlue.java

```diff
@@ -442,30 +481,43 @@ private List<PickleStepDefinitionMatch> stepDefinitionMatches(URI uri, Step step
     }
 
     void removeScenarioScopedGlue() {
-        stepDefinitionsByPattern.clear();
-        removeScenarioScopedGlue(beforeHooks);
-        removeScenarioScopedGlue(beforeStepHooks);
-        removeScenarioScopedGlue(afterHooks);
-        removeScenarioScopedGlue(afterStepHooks);
-        removeScenarioScopedGlue(stepDefinitions);
-        removeScenarioScopedGlue(dataTableTypeDefinitions);
-        removeScenarioScopedGlue(docStringTypeDefinitions);
-        removeScenarioScopedGlue(parameterTypeDefinitions);
-        removeScenarioScopedGlue(defaultParameterTransformers);
-        removeScenarioScopedGlue(defaultDataTableEntryTransformers);
-        removeScenarioScopedGlue(defaultDataTableCellTransformers);
-    }
-
-    private void removeScenarioScopedGlue(Iterable<?> glues) {
+        boolean dirty = false;
+        dirty |= removeScenarioScopedGlue(beforeHooks);
+        dirty |= removeScenarioScopedGlue(beforeStepHooks);
+        dirty |= removeScenarioScopedGlue(afterHooks);
+        dirty |= removeScenarioScopedGlue(afterStepHooks);
+        if (removeScenarioScopedGlue(stepDefinitions)) {
+            dirty = true;
+            dirtyCache = true;
+        }
+        dirty |= removeScenarioScopedGlue(dataTableTypeDefinitions);
+        dirty |= removeScenarioScopedGlue(docStringTypeDefinitions);
+        if (removeScenarioScopedGlue(parameterTypeDefinitions)) {
+            dirty = true;
+            dirtyCache = true;
+        }
+        dirty |= removeScenarioScopedGlue(defaultParameterTransformers);
+        dirty |= removeScenarioScopedGlue(defaultDataTableEntryTransformers);
+        dirty |= removeScenarioScopedGlue(defaultDataTableCellTransformers);
+        if (dirty) {
+            stepDefinitionsByPattern.clear();
```

Comment:
Is this still necessary given that `hasScenarioScopedGlue` will also clear the step definition patterns?

Reply:
Good catch.

Given that `Runner.runPickle()` prepares the glue for each pickle:
```
        try {
            ...
            glue.prepareGlue(localeForPickle(pickle));
            ...
        } finally {
            glue.removeScenarioScopedGlue();
            ...
        }
```
then we don't need to call `stepDefinitionsByPattern.clear()` in `removeScenarioScopedGlue()`. Clearing the `stepDefinitionsByPattern` will be done in `prepareGlue()` when we run the next pickle.